js/noms/src/spec.js

```
@@ -0,0 +1,315 @@
+    this._spec = spec;
+    this._opts = opts;
+    const [protocol, databaseName] = parseDatabaseSpec(databaseSpec);
+    this._protocol = protocol;
+    this._databaseName = databaseName;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a database.
+   */
+  static forDatabase(spec: string, opts: ?SpecOptions): Spec {
+    return new Spec(spec, spec, opts);
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a dataset.
+   */
+  static forDataset(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, datasetName] = splitDatabaseSpec(spec);
+    if (!datasetRe.test(datasetName)) {
+      throw new SyntaxError(`Invalid dataset ${datasetName}, must match ${datasetRe.source}`);
+    }
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._datasetName = datasetName;
+    return specObj;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a value path.
+   */
+  static forPath(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, pathStr] = splitDatabaseSpec(spec);
+    const path = AbsolutePath.parse(pathStr);
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._path = path;
+    return specObj;
+  }
+
+  /**
+   * Returns the spec string.
+   */
+  spec(): string {
+    return this._spec;
+  }
+
+  /**
+   * Returns the database protocol, for example `"mem"` or `"http"`.
+   */
+  protocol(): string {
+    return this._protocol;
+  }
+
+  /**
+   * Returns the database name, e.g. `"demo.noms.io"` or `"localhost:8000"`.
+   */
+  databaseName(): string {
+    return this._databaseName;
+  }
+
+  /**
+   * Returns the `Database`.
+   */
+  database(): Database {
+    if (!this._database) {
+      this._database = this._createDatabase();
+    }
+    return this._database;
+  }
+
+  /**
+   * Returns the dataset name, for example `"sf-crime"`.
+   *
+   * Throws a `TypeError` if this spec does not have an associated value.
+   */
+  datasetName(): string {
```

Why throw an error rather than just return nil?